src/Spanner/Result.php

```diff
@@ -0,0 +1,326 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner;
+
+use Google\Cloud\Core\Exception\ServiceException;
+use Google\Cloud\Spanner\Session\Session;
+use Google\Cloud\Spanner\Session\SessionPoolInterface;
+
+/**
+ * Represent a Google Cloud Spanner lookup result (either read or executeSql).
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Spanner\SpannerClient;
+ *
+ * $spanner = new SpannerClient();
+ *
+ * $database = $spanner->connect('my-instance', 'my-database');
+ * $result = $database->execute('SELECT * FROM Posts');
+ * ```
+ *
+ * @see https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.ResultSet ResultSet
+ */
+class Result implements \IteratorAggregate
+{
+    /**
+     * @var array|null
+     */
+    private $cachedValues;
+
+    /**
+     * @var array
+     */
+    private $columns = [];
+
+    /**
+     * @var ValueMapper
+     */
+    private $mapper;
+
+    /**
+     * @var array|null
+     */
+    private $metadata;
+
+    /**
+     * @var Operation
+     */
+    private $operation;
+
+    /**
+     * @var string|null
+     */
+    private $resumeToken;
+
+    /**
+     * @var Session
+     */
+    private $session;
+
+    /**
+     * @var Snapshot|null
+     */
+    private $snapshot;
+
+    /**
+     * @var array|null
+     */
+    private $stats;
+
+    /**
+     * @var Transaction|null
+     */
+    private $transaction;
+
+    /**
+     * @var string
+     */
+    private $transactionContext;
+
+    /**
+     * @param Operation $operation Runs operations against Google Cloud Spanner.
+     * @param Session $session The session used for any operations executed.
+     * @param \Generator $resultGenerator Reads rows from Google Cloud Spanner.
+     * @param string $transactionContext The transaction's context.
+     * @param ValueMapper $mapper Maps values.
+     */
+    public function __construct(
+        Operation $operation,
+        Session $session,
+        callable $call,
+        $transactionContext,
+        ValueMapper $mapper
+    ) {
+        $this->operation = $operation;
+        $this->session = $session;
+        $this->call = $call;
+        $this->transactionContext = $transactionContext;
+        $this->mapper = $mapper;
+    }
+
+    /**
+     * Return the formatted and decoded rows.
+     *
+     * If the stream is interrupted an attempt will be made to resume.
+     *
+     * Example:
+     * ```
+     * $rows = $result->rows();
+     * ```
+     *
+     * @return \Generator
+     */
+    public function rows()
+    {
+        $call = $this->call;
+
+        try {
+            foreach ($this->getRows($call()) as $row) {
+                yield $row;
+            }
+        } catch (ServiceException $ex) {
+            if (!$this->resumeToken) {
+                throw $ex;
+            }
+
+            // If we have a token, attempt to resume
+            foreach ($this->getRows($call($this->resumeToken)) as $row) {
```

Comment:
We should retry till deadline expires. Also we should only retry if the exception is retryable and not for all exceptions.